cf-check/validate.c

```diff
@@ -0,0 +1,515 @@
+                ValidationError(
+                    state, "Missing address entry for '%s'", address);
+            }
+            else if (!StringSafeEqual(hostkey, lookup))
+            {
+                ValidationError(
+                    state,
+                    "Bad hostkey->address->hostkey reverse lookup '%s' != '%s'",
+                    hostkey,
+                    lookup);
+            }
+        }
+    }
+    {
+        MapIterator iter = MapIteratorInit(address_to_hostkey->impl);
+        MapKeyValue *current_item;
+        while ((current_item = MapIteratorNext(&iter)) != NULL)
+        {
+            const char *address = current_item->key;
+            if (!ValidateAddress(state, address))
+            {
+                continue;
+            }
+            const char *hostkey = current_item->value;
+            if (!ValidateHostkey(state, hostkey))
+            {
+                continue;
+            }
+            const char *lookup = StringMapGet(hostkey_to_address, hostkey);
+            if (lookup == NULL)
+            {
+                ValidationError(
+                    state, "Missing hostkey entry for '%s'", hostkey);
+            }
+            else if (!StringSafeEqual(address, lookup))
+            {
+                ValidationError(
+                    state,
+                    "Bad address->hostkey->address reverse lookup '%s' != '%s'",
+                    address,
+                    lookup);
+            }
+        }
+    }
+    {
+        StringSetIterator iter =
+            StringSetIteratorInit(quality_incoming_hostkeys);
+        const char *hostkey;
+        while ((hostkey = StringSetIteratorNext(&iter)) != NULL)
+        {
+            if (StringMapGet(hostkey_to_address, hostkey) == NULL)
+            {
+                ValidationError(
+                    state,
+                    "Missing hostkey from quality in entry '%s'",
+                    hostkey);
+            }
+        }
+    }
+    {
+        StringSetIterator iter =
+            StringSetIteratorInit(quality_outgoing_hostkeys);
+        const char *hostkey;
+        while ((hostkey = StringSetIteratorNext(&iter)) != NULL)
+        {
+            if (StringMapGet(hostkey_to_address, hostkey) == NULL)
+            {
+                ValidationError(
+                    state,
+                    "Missing hostkey from quality out entry '%s'",
+                    hostkey);
+            }
+        }
+    }
+}
+
+static void ValidateState(ValidatorState *state)
+{
+    assert(state != NULL);
+
+    switch (state->mode)
+    {
+    case CF_CHECK_VALIDATE_LASTSEEN:
+        ValidateStateLastseen(state);
+    case CF_CHECK_VALIDATE_UNKNOWN:
+        break;
+    default:
+        debug_abort_if_reached();
+        break;
+    }
+}
+
+int CFCheck_Validate(const char *path)
+{
+    assert(path != NULL);
+
+    int rc;
+
+    MDB_env *env;
+    rc = mdb_env_create(&env);
+    if (rc)
+        return lmdb_report_error(rc);
```

Comment:
This violates at least two rules in our guidelines! :wink:

Reply:
Copy-paste.

Reply:
Fixed.